fml/synchronization/sync_switch.h

```diff
@@ -53,7 +53,7 @@ class SyncSwitch {
   void SetSwitch(bool value);
 
  private:
-  mutable std::mutex mutex_;
+  mutable std::recursive_mutex mutex_;
```

Comment:
Is there a reason this is recursive?  I'd rather not make it recursive.  I don't think anyone under DoDraw should be executing the sync switch, no?

Reply:
Replaced here with `std::recursive_mutex` is to let Scenario App Integration Tests pass，This is the test result before the replacement：https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8837385340239046353/+/u/Scenario_App_Integration_Tests/stdout?format=raw. 
The reason is that external view embedders also use `SyncSwitch`. https://github.com/flutter/engine/pull/22302
So there are two options. 
1.  Use `std::recursive_mutex` instead of `std::mutex`. This solution has a small workload but may have poor performance
2. Remove the SyncSwitch used by external view embedders. This solution may have some workload
Please let me know if you prefer the second option, my original plan was to land this PR first, and then see if we can remove the SyncSwitch in the external view embedders.

Reply:
I think we should do the second thing, remove the sync switch from the external view embedders, this lock supersedes that one.

```
Rasterizer::Draw
  -> Rasterizer::DoDraw
    -> Rasterizer::DrawToSurface
      -> FlutterPlatformViewsController::SubmitFrame
        -> SyncSwitch::Execute
```

However look at `Rasterizer::DrawLastLayerTree`, potentially that is using the sync switch since that calls `Rasterizer::DrawToSurface` outside of `Rasterizer::Draw`

```
Rasterizer::DrawLastLayerTree
  -> Rasterizer::DrawToSurface
    -> FlutterPlatformViewsController::SubmitFrame
      -> SyncSwitch::Execute
```

That means we should use the sync switch in `Rasterizer::DrawToSurface` instead of `Rasterizer::Draw`.  Instead of removing or adding a sync switch, we can think of this as just moving the existing one in FlutterPlatformViewsController up one more layer.  From the stack traces we've seen this should be a good place to have it still.

Reply:
Currently, if we are in the background, the `RasterStatus` we return is `RasterStatus::kDiscarded`.I can be sure that there is no problem, because the following code also returns `RasterStatus::kDiscarded` in `Draw`.
```
 if (discardCallback(*layer_tree.get())) {
   raster_status = RasterStatus::kDiscarded;
 } 
```
But if we move sync switch to `Rasterizer::DrawToSurface`, and in the background in iOS we retuns  `RasterStatus::kDiscarded` from `Rasterizer::DrawToSurface`, how should `Rasterizer::DrawToSurface` handle it?
https://github.com/flutter/engine/blob/master/shell/common/rasterizer.cc#L380-L388

Reply:
`RasterStatus::kDiscarded` is Added by https://github.com/flutter/engine/pull/21179.

Reply:
I'd return kDiscarded from DrawToSurface and in DoDraw expand the branch that checks return codes to early out like so:
```c++
  RasterStatus raster_status =
      DrawToSurface(*frame_timings_recorder, *layer_tree);
  if (raster_status == RasterStatus::kSuccess) {
    last_layer_tree_ = std::move(layer_tree);
  } else if (raster_status == RasterStatus::kResubmit ||
             raster_status == RasterStatus::kSkipAndRetry) {
    resubmitted_layer_tree_ = std::move(layer_tree);
    return raster_status;
  } else if (raster_status = RasterStatus::kDiscarded) {
    return raster_status;
  }
```

I think that should be good.

Reply:
I got it. Another problem is that we need to wrap a method outside `DrawToSurface` to handle sync switch (such as `DrawToSurfaceWrapper`), `DoDraw` and `DrawLastLayerTree ` should call `DrawToSurfaceWrapper` instead of `DrawToSurface` . Here I need a good method name because I am not a native speaker.

Reply:
"Wrapper" isn't that good since it doesn't tell you anything interesting about the code's semantics are, just how it is structured.

How about his:
```c++
RasterStatus DrawToSurface() {
  RasterStatus result;
  ...
  IsGpuDisabledSyncSwitch().execute(Handlers()
    .SetIfTrue([]{result = kDiscarded;})
    .SetIfFalse([]{result = DrawToSurfaceUnSafe()});
  ...
}
```

We use the "UnSafe" postfix in other code in the engine to denote the context with which it should be uses is enforced higher in the call stack.

Is that was you were asking about?

Reply:
> Is that was you were asking about?
Yes, I got it